### ISSUE TITLE
fastlane changelongs limit

### ISSUE THREAD

**Issue by IzzySoft** [LucasGGamerM/moshidon#317]
My scanner just reported:

    en-US/changelogs/103.txt has 691 chars

Please remember (for future releases) that those files are limited to 500 chars each. If they are longer, they are trucated at exactly that position, most likely mid-word. I've adjusted 103.txt on my end to prevent this, but please take care for future releases :wink:

**Comment by LucasGGamerM** [comment]
Understood! I sometimes go overboard when writing the changelogs :D

**Comment by IzzySoft** [comment]
Which is not bad per se. Suggestion: Have a `Changelog.md` for the full list of changes (always appending new releases on-top), stick to the most important ones in your per-release fastlane changelogs, and end the latter with something like `For a full list of changes, see https://github.com/LucasGGamerM/moshidon/blob/master/Changelog.md`. Then have the best of both worlds :wink:

**Comment by IzzySoft** [comment]
@LucasGGamerM As this is still open, permit me to add one question:

It seems you currently have 2 different locations for the fastlane tree. HEAD and tags 2.9.x have it at `metadata/`, the *moshinda tags at `fastlane/`. That makes it a bit tricky for us to handle. Which one will be the "real location"? Metadata have been in `fastlane/` for quite some time, so I also wonder why you move them to a non-standard (F-Droid specific) location?

**Comment by IzzySoft** [comment]
PS: yuck, it seems the v2.9.x are the upstream ones, and the other are yours. That makes it tricky. We evaluate the latest tag for cloning via `git ls-remote`, which does not know about timestamps. Also, Github has a bug with releases, where in some specific cases the latest tag is not shown there, so that API is not reliable either. Which means, for cloning we'll always get Grishka's tags, as they show the higher version (per [semantic versioning](https://semver.org/). 😢 A quite specific edge-case…

**Comment by IzzySoft** [comment]
If I knew I had included that… In the long run, an update to our framework so the tag found by the APK updater (which uses the Github API) is passed along, maybe. But that would then only work for new releases, not for "manual fastlane syncs".

Fastlane uses `git ls-remote`, then sorts the tags according to semver. So if you see an easy way to rename the upstream tags (prepending a `0.`), that would solve it – but that's not feasible at your end. Pushing up your own versions beyond upstream (e.g. 3.x.y) would solve it, too – but according to semver, that would signal a major change you probably do not want to send.

We could introduce a per-app "window" for when a tag is to be considered too old (currently, if it's older than 90d, we pull from the default branch instead) – but that again would be quite some effort for a single app or two needing it.

Not sure what's left then. Not sure if using `<UpstreamVersion>+<build identifier>` would do the trick, e.g. `2.9.7+fork.109` – but then again, that would only work if you always publish a new build each time after upstream does.

----

**TL;DR:**

Best bet is us handing in the tag pattern set; for Moshidon, currently `Version .+fork\.%c\.moshinda` – which would ignore all tags not ending on `[0-9]+.moshinda`. There are some other apps that would profit from this. I'll put that on my list, but it might take a while; until then, please bear with us that this can "run out of sync" – and will, whenever there's any upstream tag newer than 90d when you release.

Apologies for the noise: there are so many edge-cases when you rewrite a framework, and some of them only show up after you merged…

**Comment by IzzySoft** [comment]
@LucasGGamerM was easier than I thought, should be solved now:

```
$ iod repo clonerepo org.joinmastodon.android.moshinda
- Cloning upstream repo …
Tag (old method): 'v2.9.7'
Tag (new method): '2.3.0+fork.110.moshinda'
```

So it should pick the correct tag now, phew… 😌 (and of course that debug output is meanwhile removed 🤪)

Btw, testing the effect of the updated routine against a bunch of other apps showed there were 3 more affected (which we didn't notice as they had no updates since). Glad that's solved now!
